pyphare/pyphare/pharesee/hierarchy.py

```diff
@@ -409,6 +409,25 @@ def levelNbr(self, time):
     def levelNbrs(self, time):
         return list(self.levels(time).keys())
 
+    def is_homogeneous(self):
+        """
+        return True if all patches of all levels at all times
+        have the same patch data quantities
+        """
+        qties = self._quantities()
+        it_is = True
+        for time, levels in self.time_hier.items():
+            for ilvl, lvl in levels.items():
+                for patch in lvl.patches:
+                    it_is &= qties == list(patch.patch_datas.keys())
+        return it_is
+
+    def _quantities(self):
+        return list(self.level(0).patches[0].patch_datas.keys())
+
+    def quantities(self):
+        assert self.is_homogeneous()
```

Comment:
this doesn't exist for "python -O" executions

maybe not an issue

Reply:
I don't understand what you mean

Reply:
should it be a "raise RuntimeError" or an assert. as the assert is disabled when running with "python3 -O script.py"

Reply:
ok I didn't know that. I'll change it to a raise RunTimeError then